### ISSUE TITLE
Memory leak with addCurve

### ISSUE THREAD

**Issue by vallsv** [silx-kit/silx#3372]
Intensive use of `addCurve` looks to leak memory.

Here is a script from @t20100 reproducing the problem.
```
    def run(self):
        """Method implementing thread loop that updates the plot"""
        while self.running:
            time.sleep(0.1)

            # Run plot update asynchronously
            x = numpy.arange(1000)
            data = numpy.random.random(4000).reshape(4, 1000)
            for index in range(4):
                concurrent.submitToQtMainThread(
                    self.plot1d.addCurve,
                    x,
                    data[index],
                    resetzoom=False,
                    legend=str(index))
            #concurrent.submitToQtMainThread(self.plot1d.clear)
```

Here is the result in Flint after few times (memory (KB) per time is second)

![image](https://user-images.githubusercontent.com/7579321/106768979-c854be80-663c-11eb-97cf-c0089e2d9ab4.png)

**Comment by vasole** [comment]
Have you tried to change backend (Matplotlib and OpenGL) to see if the problem is at the backend side or at our side?

**Comment by vallsv** [comment]
Yes it's the same. With OpenGL the leak is smaller, but there is still something.

**Comment by t20100** [comment]
Here is a minimalistic sample code to reproduce:
```python
import numpy

from silx.gui import qt
from silx.gui.plot import PlotWidget

app = qt.QApplication([])
plot = PlotWidget()
plot.show()

x = numpy.arange(1000)
y = numpy.random.random(1000)
legend = 'curve'

plot.addCurve(x, y, resetzoom=False, legend=legend)
curve = plot.getItems()[0]

def timeout():
    curve.setData(x, y)

timer = qt.QTimer()
timer.timeout.connect(timeout)
timer.start(100)

app.exec_()
```

**Comment by cpascual** [comment]
Hi, I've tried the "minimalistic" script and I do not see any memory leak.

I monitored the process memory with:

```console
taurus trend 'eval:@p=psutil.Process($PID)/p.memory_info().rss/1024' 
```
... where $PID is the process ID of the silx plot script

Note: the script does not change the values of x,y (just sets the same again and again). Is that on purpose?
just in case I tried recomputing x,y inside the timeout function so that new values are used. No mem leak either.

I also tested the equivalent code with pyqtgraph with same negative results.

Here is the info on my system (fresh conda environment running on a debian buster):

```
$ conda list                                                                                                 (silx) 
# packages in environment at /home/cpascual/miniconda/envs/silx:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
backcall                  0.2.0              pyh9f0ad1d_0    conda-forge
backports                 1.0                        py_2    conda-forge
backports.functools_lru_cache 1.6.1                      py_0    conda-forge
c-ares                    1.17.1               h36c2ea0_0    conda-forge
ca-certificates           2020.12.5            ha878542_0    conda-forge
cached-property           1.5.1                      py_0    conda-forge
certifi                   2020.12.5        py37h89c1867_1    conda-forge
cycler                    0.10.0                     py_2    conda-forge
dbus                      1.13.6               hfdff14a_1    conda-forge
decorator                 4.4.2                      py_0    conda-forge
expat                     2.2.10               h9c3ff4c_0    conda-forge
fabio                     0.11.0           py37h902c9e0_1    conda-forge
fontconfig                2.13.1            hba837de_1004    conda-forge
freetype                  2.10.4               h0708190_1    conda-forge
gettext                   0.19.8.1          h0b5b191_1005    conda-forge
glib                      2.66.5               hcd2ae1e_2    conda-forge
glib-tools                2.66.5               hcd2ae1e_2    conda-forge
glymur                    0.9.3              pyhd8ed1ab_0    conda-forge
gst-plugins-base          1.14.0               hbbd80ab_1  
gstreamer                 1.14.0               h28cd5cc_2  
h5py                      3.1.0           nompi_py37h1e651dc_100    conda-forge
hdf5                      1.10.6          nompi_h6a2412b_1114    conda-forge
icu                       58.2              hf484d3e_1000    conda-forge
ipykernel                 5.1.0              pyh24bf2e0_0    conda-forge
ipython                   7.20.0           py37h888b3d9_1    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
jedi                      0.18.0           py37h89c1867_2    conda-forge
jpeg                      9d                   h36c2ea0_0    conda-forge
jupyter_client            6.1.11             pyhd8ed1ab_1    conda-forge
jupyter_core              4.7.1            py37h89c1867_0    conda-forge
kiwisolver                1.3.1            py37h2527ec5_1    conda-forge
krb5                      1.17.2               h926e7f8_0    conda-forge
lcms2                     2.11                 hcbb858e_1    conda-forge
ld_impl_linux-64          2.35.1               hea4e1c9_2    conda-forge
libblas                   3.9.0                8_openblas    conda-forge
libcblas                  3.9.0                8_openblas    conda-forge
libcurl                   7.71.1               hcdd3856_8    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 h516909a_1    conda-forge
libffi                    3.3                  h58526e2_2    conda-forge
libgcc-ng                 9.3.0               h2828fa1_18    conda-forge
libgfortran-ng            9.3.0               hff62375_18    conda-forge
libgfortran5              9.3.0               hff62375_18    conda-forge
libglib                   2.66.5               h164308a_2    conda-forge
libgomp                   9.3.0               h2828fa1_18    conda-forge
libiconv                  1.16                 h516909a_0    conda-forge
liblapack                 3.9.0                8_openblas    conda-forge
libnghttp2                1.43.0               h812cca2_0    conda-forge
libopenblas               0.3.12          pthreads_h4812303_1    conda-forge
libpng                    1.6.37               h21135ba_2    conda-forge
libsodium                 1.0.18               h36c2ea0_1    conda-forge
libssh2                   1.9.0                hab1572f_5    conda-forge
libstdcxx-ng              9.3.0               h6de172a_18    conda-forge
libtiff                   4.2.0                hdc55705_0    conda-forge
libuuid                   2.32.1            h7f98852_1000    conda-forge
libwebp-base              1.2.0                h7f98852_0    conda-forge
libxcb                    1.13              h7f98852_1003    conda-forge
libxml2                   2.9.10               hb55368b_3  
libxslt                   1.1.34               hc22bd24_0  
lxml                      4.6.2            py37h77fd288_1    conda-forge
lz4-c                     1.9.3                h9c3ff4c_0    conda-forge
mako                      1.1.4              pyh44b312d_0    conda-forge
markupsafe                1.1.1            py37h5e8e339_3    conda-forge
matplotlib                3.3.4            py37h89c1867_0    conda-forge
matplotlib-base           3.3.4            py37h0c9df89_0    conda-forge
ncurses                   6.2                  h58526e2_4    conda-forge
numpy                     1.20.0           py37haa41c4c_0    conda-forge
olefile                   0.46               pyh9f0ad1d_1    conda-forge
openjpeg                  2.4.0                hf7af979_0    conda-forge
openssl                   1.1.1i               h7f98852_0    conda-forge
parso                     0.8.1              pyhd8ed1ab_0    conda-forge
pcre                      8.44                 he1b5a44_0    conda-forge
pexpect                   4.8.0              pyh9f0ad1d_2    conda-forge
pickleshare               0.7.5                   py_1003    conda-forge
pillow                    8.1.0            py37he6b4880_1    conda-forge
pip                       21.0.1             pyhd8ed1ab_0    conda-forge
prompt-toolkit            3.0.14             pyha770c72_0    conda-forge
pthread-stubs             0.4               h36c2ea0_1001    conda-forge
ptyprocess                0.7.0              pyhd3deb0d_0    conda-forge
pygments                  2.7.4              pyhd8ed1ab_0    conda-forge
pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
pyqt                      5.9.2            py37hcca6a23_4    conda-forge
pyqtgraph                 0.11.1             pyhd3deb0d_0    conda-forge
python                    3.7.9           hffdb5ce_0_cpython    conda-forge
python-dateutil           2.8.1                      py_0    conda-forge
python_abi                3.7                     1_cp37m    conda-forge
pyzmq                     22.0.1           py37h499b945_0    conda-forge
qt                        5.9.7                h5867ecd_1  
qtconsole                 5.0.2              pyhd8ed1ab_0    conda-forge
qtpy                      1.9.0                      py_0    conda-forge
readline                  8.0                  he28a2e2_2    conda-forge
scipy                     1.6.0            py37h14a347d_0    conda-forge
setuptools                49.6.0           py37h89c1867_3    conda-forge
silx                      0.14.0               hd8ed1ab_0    conda-forge
silx-base                 0.14.0           py37hdc94413_0    conda-forge
sip                       4.19.8           py37hf484d3e_0  
six                       1.15.0             pyh9f0ad1d_0    conda-forge
sqlite                    3.34.0               h74cdb3f_0    conda-forge
tk                        8.6.10               h21135ba_1    conda-forge
tornado                   6.1              py37h5e8e339_1    conda-forge
traitlets                 5.0.5                      py_0    conda-forge
wcwidth                   0.2.5              pyh9f0ad1d_2    conda-forge
wheel                     0.36.2             pyhd3deb0d_0    conda-forge
xorg-libxau               1.0.9                h7f98852_0    conda-forge
xorg-libxdmcp             1.1.3                h7f98852_0    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
zeromq                    4.3.4                h9c3ff4c_0    conda-forge
zlib                      1.2.11            h516909a_1010    conda-forge
zstd                      1.4.8                ha95c52a_1    conda-forge
```

**Comment by t20100** [comment]
Thanks for the testing!

> Note: the script does not change the values of x,y (just sets the same again and again). Is that on purpose?
Yes idea was to check if there is any memory leak in the machinery, whether data is different or not should not change the behavior.

I'm using `mprof run` from [memory-profiler](https://pypi.org/project/memory-profiler/) (which looks to use `psutil`), and I see a small increase over time:

![Figure_1](https://user-images.githubusercontent.com/9449698/106894885-d1eb2e80-66ef-11eb-9400-6b363365e1ae.png)

But I'm not sure I'm looking at the right thing, or maybe memory usage stop increasing after a while...

**Comment by cpascual** [comment]
Hi, here is what I get with mprof run + mprof plot using the "minimal" script (I have only run it for ~2min, but I think it is safe to say that there is no memleak)
![Figure_1](https://user-images.githubusercontent.com/4115224/106912279-29df6080-6703-11eb-9902-ccd57f0b54bc.png)

For comparison, this is what I get from doing the same with pyqtgraph
![Figure_2](https://user-images.githubusercontent.com/4115224/106911356-5646ad00-6702-11eb-97bf-87ed13f6673c.png)

**Comment by cpascual** [comment]
note: in my tests I avoiding doing any other thing with the PC at the same time, to avoid "contaminating" the results with moving windows, etc.

**Comment by t20100** [comment]
OK, it looks like what I would expect.

I run it without doing anything else and here is the result, it stabilises after a while:

![Figure_1](https://user-images.githubusercontent.com/9449698/106918687-b7be4a00-6709-11eb-8e44-134473302ca9.png)

BTW, I'm on macos so things are probably different underneath.

**Comment by vallsv** [comment]
Sounds like you don't check the same thing as me.

**Comment by t20100** [comment]
I tried to simplified the pattern as much as I can... but probably too much as I got lured by the increase when doing quick tests.

**Comment by vallsv** [comment]
Here is mine.

If you use a single curve you don't have the leak.

Spend so much time for a so simple sample that i can't really believe this result.

```
import sys
import os
import numpy
from silx.gui import qt
from silx.gui.plot.PlotWindow import Plot1D


def create_window():
    plot = Plot1D(None)

    x = numpy.zeros((1,))

    def update():
        plot.addCurve(x, x, legend="a")
        plot.addCurve(x, x, legend="b")

    timer = qt.QTimer(plot)
    timer.setInterval(1)
    timer.timeout.connect(update)
    timer.start()

    return plot


def main():
    qapp = qt.QApplication([])
    window = create_window()
    window.setWindowTitle(f"{os.getpid()}")
    window.setVisible(True)
    sys.exit(qapp.exec_())


if __name__ == "__main__":
    main()
```

**Comment by vallsv** [comment]
Found that `matplotlib.cbook.CallbackRegistry.disconnect` is never called.

As result the dict `self._func_cid_map[s]` grow up always, storing weakref.

But most probably the problem come from something else.

**Comment by vallsv** [comment]
Here is a pure MPL example.
I think it is more or less what we do in silx.
```
import sys
import os
from PyQt5 import Qt

from matplotlib.backends.backend_qt5agg import FigureCanvasQTAgg
from matplotlib.figure import Figure
from matplotlib.container import Container


class PlotCanvas(FigureCanvasQTAgg):

    def __init__(self, parent=None, width=5, height=4, dpi=100):
        fig = Figure(figsize=(width, height), dpi=dpi)
        self.axes = fig.add_subplot(111)

        FigureCanvasQTAgg.__init__(self, fig)
        self.setParent(parent)

        self.l1 = None
        self.l2 = None

        timer = Qt.QTimer(self)
        timer.setInterval(1)
        timer.timeout.connect(self.tick)
        timer.start()

    def tick(self):
        if self.l1 is not None:
            self.l1.remove()
        if self.l2 is not None:
            self.l2.remove()

        self.l1 = Container([self.axes.plot([1,2], [1,2])])
        self.l2 = Container([self.axes.plot([2,1], [1,2])])

        self.draw()

def main():
    app = Qt.QApplication(sys.argv)
    window = PlotCanvas()
    window.setWindowTitle(f"{os.getpid()}")
    window.setVisible(True)
    return app.exec_()

if __name__ == "__main__":
    result = main()
    sys.exit(result)
```

**Comment by vallsv** [comment]
It looks like it is fine if we use another container, that's really matter of timing.

```
class MyContainer:

    def __init__(self, axes, artists):
        self.axes = axes
        self.artists = artists

    def remove(self):
        artists = self.artists # without this 2 lines, it still fails
        self.artists = None  # need python expert here o_O or a sleep
        for a in artists:
            a.remove()
```

**Comment by vallsv** [comment]
Found a fix. Sounds like a pure mpl bug.

**Comment by vallsv** [comment]
If you are about to release silx soon, would you like a kind of monkey patch before the fix in matplotlib?

I am afraid i will have to do it anyway for Flint, so i can patch silx instead.

What do you think?

**Comment by t20100** [comment]
I'd prefer a workaround on our side rather than a monkey patch... but in this case it's probably complicated.

Looking at the history of `CallbackRegistry`, it didn't changed much in the last few years... so I expect we can safely monkey patch it for a wide-enough range of versions without testing with each individual version.

**Comment by vallsv** [comment]
We did that in Flint: https://gitlab.esrf.fr/bliss/bliss/-/merge_requests/3469. It's fine for me.

Also, up to you if you want to close the issue, or let it open until it is not fixed in matplotlib.

**Comment by t20100** [comment]
OK, then unless someone else thinks that we need to fix it from silx, I'd leave it as it is since it should be fixed in `matplotlib`.

**Comment by vallsv** [comment]
The patch was released with matplotlib 3.4.0 RC1.
Then i close this issue.
